### PR TITLE
accounts, core, crypto, internal: use canonical sigs internally

### PR DESCRIPTION
**Pull request by karalabe**
Ethereum uses the `secp256k1` curve cryptography, which has it's own "canonical" signature representation. For the scope of this PR, we only care about the final byte (value of `V`) being `0` or `1` canonically. Both our `crypto` library as well as our `AccountManager` provide methods for generating these signatures.

However, Ethereum transactions themselves are **not** signed using the canonical representation, rather the yellow paper specifies adding `27` to the `V` value. This cause both `crypto` and `AccountManager` packages gaining additional methods `SingEthereum` and `SignEthereumWithPassphrase`. The problem is that anyone using our code as a library now needs to be aware of the crypto differences to use the methods properly.

This issue is further complicated by EIP158 which introduced chain IDs, which are yet another different schemes about how to transform the `V` value to do chain replay protections. Opposed to the plain and "Ethereum" style signatures, these are hidden from the user even at a library level. However now there's yet again a mental note people need to be aware of.

---

To address this increasing signature complexity, this PR discards all notion of "different" signature types at the library level. Both the `crypto` and `accounts` package is reduced to only be able to produce plain canonical `secp256k1` signatures. This makes the crpyto APIs much cleaner, simpler and harder to abuse.

To handle the Ethereum yellow paper or Ethereum EIP158 style signature transforms, all the details have been moved out to the API edges and only executed when a particular signature format is needed. E.g. the Frontier and Homestead signers take a canonical signature and transform them into the canonical format, whereas the EIP158 signer takes again the canonical signature and transforms it according to our chain ID. This allows users to call a single Sign method and have it behave correctly irrelevant where the signature is used.

Similarly the RPC API needed the plain -> yellow papers transforms implemented there too.

### CODE REVIEW

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -1113,12 +1117,19 @@ func (s *PublicTransactionPoolAPI) SendRawTransaction(ctx context.Context, encod
 // Sign calculates an ECDSA signature for:
 // keccack256("\x19Ethereum Signed Message:\n" + len(message) + message).
 //
+// Note, the produced signature conforms to the Ethereum yellow paper spec where
+// the V parameter must be 27 or 28 irrelevant of any ChainID features.
+//
```
Comment (fjl) [inline internal/ethapi/api.go]: Instead of using this sentence ("conforms to the Ethereum yellow paper spec...") please document that the format of the resulting signature is `r, s, v where v is 0 or 1`. There is no need to mention ChainID here.

File: accounts/account_manager.go [hunk 2]
```diff
@@ -136,42 +136,33 @@ func (am *Manager) DeleteAccount(a Account, passphrase string) error {
...
+	return crypto.Sign(hash, unlockedKey.PrivateKey)
 }
 
-// SignWithPassphrase signs hash if the private key matching the given
-// address can be decrypted with the given passphrase.
+// SignWithPassphrase signs hash if the private key matching the given address
+// can be decrypted with the given passphrase. The produced signature is in the
+// canonical secp256k1 format (V = 0 or 1), which can be transformed into the
+// proper Ethereum signature accoring to the yellow paper within the transaction
+// signer.
```
Comment (fjl) [inline accounts/account_manager.go]: Please remove the part where it says how the signature could be transformed by the caller. Any caller can do any transformation on the result. "the transaction signer" is one of the callers, but I'm not sure that people reading the docs on godoc would know where/what that is.

File: internal/ethapi/api.go [hunk 3]
```diff
@@ -321,6 +324,9 @@ func (s *PrivateAccountAPI) Sign(ctx context.Context, message string, addr commo
 // hash = keccak256("\x19Ethereum Signed Message:\n"${message length}${message})
 // addr = ecrecover(hash, signature)
 //
+// Note, the signature must conform to the Ethereum yellow paper spec where the V
+// parameter must be 27 or 28 irrelevant of any ChainID features.
```
Comment (fjl) [inline internal/ethapi/api.go]: No need to mention ChainID. It's sufficient to say that V must be 27 or 28 here for legacy reasons.

File: internal/ethapi/api.go [hunk 4]
```diff
@@ -312,6 +314,7 @@ func (s *PrivateAccountAPI) Sign(ctx context.Context, message string, addr commo
 	if err != nil {
 		return "0x", err
 	}
+	signature[64] += 27 // SignWithPassphrase uses canonical secp256k1 signatures (v = 0 or 1), transform to yellow paper
```
Comment (fjl) [inline internal/ethapi/api.go]: comment could be shorter

File: crypto/crypto.go [hunk 5]
```diff
@@ -181,7 +184,7 @@ func ValidateSignatureValues(v byte, r, s *big.Int, homestead bool) bool {
 	if s.Cmp(secp256k1.N) >= 0 {
 		return false
 	}
-	if r.Cmp(secp256k1.N) < 0 && (vint == 27 || vint == 28) {
+	if r.Cmp(secp256k1.N) < 0 && (vint == 0 || vint == 1) {
 		return true
 	} else {
 		return false
```
Comment (fjl) [inline crypto/crypto.go]: This can be

```go
return r.Cmp(secp256k1.N) < 0 && s.Cmp(secp256k1.N) < 0 && (v == 0 || v == 1)
```

Please move changes around ValidateSignatureValues into their own commit.

File: accounts/account_manager.go (hunk 2 above)
Comment (fjl) [inline accounts/account_manager.go]: Maybe also add the remark about `R, S, V` format here.

**Comment by karalabe** [comment]
@fjl I've fixed up all issues in the first commit and also moved anything unrelated to ValidateSignatureValues into there. The second commit remained only the validation updates. PTAL

**Comment by fjl** [comment]
It looks better now, but I still stand by my comment that there is no such thing as "canonical secp256k1 format".

The format of a signature is independent of the curve. An ECDSA
signature is a tuple of two numbers, `R` and `S`. For secp25k1 signatures, the additional `V` value is required to recover the correct public key. It is not required for plain signature verification where the
expected public key is known.

The signature format we use everywhere is just `R, S, V` concatenated. This format is not standard or canonical, it's just what we use. Signatures are usually encoded using ASN.1 DER, but we don't do that. Bitcoin has a 'compact' signature format, which is `V, R, S` with V + 27 for [arbitrary reasons (http://bitcoin.stackexchange.com/a/5089).

In summary, please remove the phrase "canonical secp256k1 format" from comments. If you want to document the format, you can say "R, S, V format where V is 0 or 1". That's unambiguous and everyone
will know what you mean. You can also define the format somewhere (e.g. package crypto docs) and then refer to it.

**Comment by karalabe** [comment]
PTAL. No, I'm not going to flatten that last commit into the first two because it would need to split up in between the two commits to avoid rebase conflicts.
